metaflow/sidecar.py

```diff
@@ -48,109 +49,184 @@ def poll(self, timeout):
-            self.emit_msg(msg)
-        except:
-            pass
-
-    def emit_msg(self, msg):
-        msg_ser = msg.serialize().encode("utf-8")
+            if not is_context_send:
+                # If we need to send context, we try that first
+                if self._send_context_remaining_tries == -1:
+                    # We could not send the context so we don't try to send this out;
+                    # restart sidecar so use the PipeUnavailableError caught below
+                    raise PipeUnavailableError()
+                elif self._send_context_remaining_tries > 0:
+                    self._send_context()
+                if self._send_context_remaining_tries == 0:
+                    self._emit_msg(msg)
+                    self._prev_message_error = False
+                    return True
+            else:
+                self._emit_msg(msg)
+                self._prev_message_error = False
+                return True
+            return False
+        except MsgTimeoutError:
+            # drop message, do not retry on timeout
+            self._logger("Unable to send message due to timeout")
+            self._prev_message_error = True
+        except Exception as ex:
+            if isinstance(ex, (PipeUnavailableError, BrokenPipeError)):
+                self._logger("Restarting sidecar due to broken/unavailable pipe")
+                self._send_context_remaining_tries = (
+                    CONTEXT_RETRY_TIMES if self._context is not None else 0
+                )
+                self.start()
+                # For context sending, we don't retry since we already did it in
+                # start
+                if is_context_send:
+                    return
+            else:
+                self._prev_message_error = True
+            if retries > 0:
+                self._logger("Retrying msg send to sidecar (due to %s)" % repr(ex))
+                return self._send_internal(msg, retries - 1, is_context_send)
+            else:
+                self._logger(
+                    "Error sending log message (exhausted retries): %s" % repr(ex)
+                )
+        return False
+
+    def _send_context(self):
+        if self._send_context_remaining_tries > 0:
+            # If we don't succeed in sending the context, we will try again
+            # next time.
+            if self._send_internal(
+                Message(MessageTypes.START, self._context), is_context_send=True
+            ):
+                self._send_context_remaining_tries = 0
+            else:
+                self._send_context_remaining_tries -= 1
+                if self._send_context_remaining_tries == 0:
+                    # Mark as "failed after try"
+                    self._send_context_remaining_tries = -1
+
+    def _emit_msg(self, msg):
+        # If the previous message had an error, we want to prepend a "\n" to this message
+        # to maximize the chance of this message being valid (for example, if the
```

Comment:
not sure I follow the reasoning (as well as why this is needed)? This seems like an issue with constructing the payload. Can't the logger/debugger implementations take care of this?